### PR TITLE
experimental GTK+4 support

### PR DESCRIPTION
**Pull request by rkitover**
Add configure support to detect GTK+4 (which is 3.90+) and set the
defines appropriately, including __WXGTK4__.

Use this define to compile out a few bits of code that are incompatible,
particularly some gdk functions, GTK+4 recommends using backend specific
functions for this, but I don't know how to do that, and this is just a
first pass to get things to compile.

### CODE REVIEW

**Comment by rkitover** [comment]
I forgot to update the bakefile related stuff, sorry, will do that later today.

**Comment by rkitover** [comment]
So I updated all the bakefile related stuff, and now there are some more compile errors to fix in the sources that got added.

I don't know why the xcode travis build is failing, it looks like it wants a program that is set as `SETFILE` in the makefiles by configure, but it's not being set for some reason, may not be related to this PR.

**Comment by rkitover** [comment]
Ok so making the bakefiles work actually produces the dozens or hundreds of GTK API incompatibilities, which I am slowly fixing, can't promise that I will finish doing that though, may just give up.

**Comment by vadz** [comment]
Thanks for looking at this! It's really disheartening to hear that there hundreds of API incompatibilities in GTK+ 4, I had a look at their [list of changes](https://developer.gnome.org/gtk4/stable/ch25s02.html) and while it seemed pretty bad, I didn't realize it was that bad... Unfortunately I haven't even started looking at GTK+ 4 yet, so I really can't do anything else than wishing you best of luck at this stage.

**Comment by paulcor** [comment]
VZ wrote:
> Thanks for looking at this! It's really disheartening to hear that
> there hundreds of API incompatibilities in GTK+ 4

You may have forgotten that deprecation warnings are disabled by
default for the GTK+3 build (because there are so many it's hard to
spot other warnings). There are currently 289 warnings, representing 82
different deprecations. `GtkStock` is by far the largest problem, 138
warnings are due to that alone.

> Unfortunately I haven't even started looking at GTK+ 4 yet

I recently started to look at it. There are going to be some Problems.
Significant Problems.

**Comment by vadz** [comment]
Damn :-( I really hoped they would calm down with API breaking changes as GTK+ matures more and more, but things seem to be going exactly in the opposite direction.

Of course, it could also be spun as an advertisement for wxWidgets: "Use our API and you won't have to rewrite your application every couple of years!". Provided that we can find resources to add GTK+ 4 support, that is...

**Comment by paulcor** [comment]
Rafael Kitover wrote:
> Ok so making the bakefiles work actually produces the dozens or
> hundreds of GTK API incompatibilities, which I am slowly fixing,
> can't promise that I will finish doing that though, may just give up.

I think it would be better to submit separate fixes for individual
deprecation problems in the GTK+3 build. That will be much easier to
deal with than one giant patch changing many things at once, and can be
built and tested without needing GTK+3.9x.

**Comment by rkitover** [comment]
@paulcor so how about I submit separate PRs for the following:

- autoconf support
- bakefile support
- each separate API fix/workaround/etc.

**Comment by vadz** [comment]
FWIW I'm fine with using a single PR for this, splitting into independent commits is good enough IMHO. But the changes definitely should be split into separate commits, starting with the the separate fixes and with the build system changes coming at the end.

TIA!

**Comment by rkitover** [comment]
@vadz @paulcor 

I've cleaned everything up and split it up into separate commits, can you guys take a look please?

**Comment by vadz** [comment]
Thanks for the work! I'll have a more detailed look soon (unless Paul beats me to it), but Travis failures need to be fixed in any case.

File: src/unix/utilsx11.cpp [hunk 1]
```diff
@@ -2680,7 +2680,7 @@ bool
 wxDoLaunchDefaultBrowser(const wxLaunchBrowserParams& params)
 {
 #ifdef __WXGTK__
-#if GTK_CHECK_VERSION(3,90,0)
+#ifdef __WXGTK3__
```
Comment (vadz) [inline src/unix/utilsx11.cpp]: We can't remove the version check completely, this function is only available in 3.22.

We also presumably will need a run-time check if we use it with GTK+ 3 too.

File: include/wx/gtk/private/addremovectrl.h [hunk 2]
```diff
@@ -38,7 +38,9 @@ class wxAddRemoveImpl : public wxAddRemoveImplBase
         GtkToolbar* const toolbar = m_tbar->GTKGetToolbar();
         GtkStyleContext* context = gtk_widget_get_style_context(GTK_WIDGET(toolbar));
         gtk_style_context_add_class(context, GTK_STYLE_CLASS_INLINE_TOOLBAR);
+#ifndef __WXGTK4__
         gtk_style_context_set_junction_sides(context, GTK_JUNCTION_TOP);
```
Comment (vadz) [inline include/wx/gtk/private/addremovectrl.h]: Do things still look acceptably without this? Could you please check the "add/remove" dialog in the dialogs sample?

It's important to fix compilation, of course, but I'd like to avoid breaking the run-time appearance too...

File: build/aclocal/bakefile.m4 [hunk 3]
```diff
@@ -745,7 +745,7 @@ AC_DEFUN([AC_BAKEFILE],
     AC_SUBST(OBJCXXFLAGS)
 
 
-    BAKEFILE_BAKEFILE_M4_VERSION="0.2.11"
+    BAKEFILE_BAKEFILE_M4_VERSION="0.2.10"
```
Comment (vadz) [inline build/aclocal/bakefile.m4]: This doesn't look right.

File: build/bakefiles/common.bkl [hunk 4]
```diff
@@ -2,7 +2,7 @@
 
 <makefile>
 
-    <requires version="0.2.7"/>
+    <requires version="0.2.10"/>
```
Comment (vadz) [inline build/bakefiles/common.bkl]: Is this really necessary? Not that it hurts anyhow, of course, but I just wonder why was this changed.

File: configure.in [hunk 5]
```diff
@@ -6487,7 +6500,7 @@ if test "$wxUSE_UIACTIONSIMULATOR" = "yes" ; then
                     AC_DEFINE(wxUSE_XTEST)
                 ],
                 [
-                    if test "$WXGTK3" = 1; then
+                    if test "$WXGTK3" = 1 -o "$WXGTK4" = 1; then
```
Comment (vadz) [inline configure.in]: It might be more clear to define `WXGTK_3_OR_4` variable once and use it rather than testing for both of those in multiple places IMO.

**Comment by vadz** [comment]
Mostly this looks good, thanks for separating this in nicely atomic commits, this made it much simpler to review the changes.

And thanks again for the changes themselves, of course!

File: build/bakefiles/common.bkl (hunk 4 above)
Comment (rkitover) [reply build/bakefiles/common.bkl]: The reason for the two changes above, is that there actually is no bakefile 0.2.11, the latest version of the legacy branch is 0.2.10, and the version check is broken so bakefile thinks that 0.2.10 is 0.2.1 and is below 0.2.7.

Comment (vadz) [reply build/bakefiles/common.bkl]: This is indeed annoying and confusing (sorry...), but while the 0.2.11 release hasn't been done yet, the [latest branch](https://github.com/vslavik/bakefile/tree/legacy-0.2-branch) definitely does use 0.2.11.

Comment (rkitover) [reply build/bakefiles/common.bkl]: Ah ok, I'll change that back then and use the git version.

File: configure.in [hunk 6]
```diff
@@ -5034,6 +5040,11 @@ else
   fi
 fi
 
+if test "$WXGTK4" = 1 ; then
+    AC_DEFINE_UNQUOTED(__WXGTK4__, 1)
+    WXGTK3=1
+    WXGTK2=1
```
Comment (paulcor) [inline configure.in]: Don't need `WXGTK2=1` here, the code immediately below will do it.

File: configure.in (hunk 5 above)
Comment (paulcor) [inline configure.in]: `WXGTK4` implies `WXGTK3` here, so this change is not needed.

File: configure.in [hunk 7]
```diff
@@ -6497,7 +6510,7 @@ if test "$wxUSE_UIACTIONSIMULATOR" = "yes" ; then
                     wxUSE_XTEST="no"
                 ]
             )
-        elif test "$WXGTK3" = 1; then
+        elif test "$WXGTK3" = 1 -o "$WXGTK4" = 1; then
```
Comment (paulcor) [inline configure.in]: `WXGTK4` implies `WXGTK3`

File: configure.in [hunk 8]
```diff
@@ -7125,7 +7138,7 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
     USE_WEBVIEW_WEBKIT2=0
     if test "$wxUSE_WEBVIEW_WEBKIT" = "yes"; then
         if test "$wxUSE_GTK" = 1; then
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1 -o "$WXGTK3" = 1; then
```
Comment (paulcor) [inline configure.in]: `WXGTK4` implies `WXGTK3`

File: configure.in [hunk 9]
```diff
@@ -7778,7 +7793,7 @@ elif test "$GXX" = yes ; then
     dnl that they're deprecated but we still have to use them to support older
     dnl toolkit versions and leaving this warning enabled prevents seeing any
     dnl other ones
-    if test "$WXGTK3" = 1 -o "$wxUSE_MAC" = 1 ; then
+    if test "$WXGTK4" = 1 -o "$WXGTK3" = 1 -o "$wxUSE_MAC" = 1 ; then
```
Comment (paulcor) [inline configure.in]: I don't think we want to disable deprecation warnings for GTK+4. At least not yet.

File: configure.in [hunk 10]
```diff
@@ -7139,7 +7152,9 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
             fi
             if test "$USE_WEBVIEW_WEBKIT2" = 0; then
                 webkitgtk=webkit-1.0
-                if test "$WXGTK3" = 1; then
+                if test "$WXGTK4" = 1; then
+                    webkitgtk=webkitgtk-4.0
+                elif test "$WXGTK3" = 1; then
```
Comment (paulcor) [inline configure.in]: Could be simplified as `webkitgtk=webkitgtk-$TOOLKIT_VERSION.0`

File: configure.in [hunk 11]
```diff
@@ -7977,7 +7992,9 @@ case "$TOOLKIT" in
     GTK)
         TOOLKIT_DESC="GTK+"
         if test "$WXGTK2" = 1; then
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1; then
+                TOOLKIT_DESC="$TOOLKIT_DESC 4"
+            elif test "$WXGTK3" = 1; then
```
Comment (paulcor) [inline configure.in]: Could simplify this as `TOOLKIT_DESC="$TOOLKIT_DESC $TOOLKIT_VERSION"`

File: configure.in [hunk 12]
```diff
@@ -5489,7 +5500,9 @@ if test "$WXGTK2" = 1; then
     if test "$wxUSE_PRINTING_ARCHITECTURE" = "yes" ; then
 
         if test "$wxUSE_GTKPRINT" = "yes" ; then
-            if test "$WXGTK3" = 1; then
+            if test "$WXGTK4" = 1; then
+                gtk_unix_print="gtk+-unix-print-4.0"
+            elif test "$WXGTK3" = 1; then
```
Comment (paulcor) [inline configure.in]: I think this could be simplified as `gtk_unix_print="gtk+-unix-print-$TOOLKIT_VERSION.0"`

File: src/unix/utilsx11.cpp (hunk 1 above)
Comment (paulcor) [inline src/unix/utilsx11.cpp]: `#ifdef __WXGTK4__`. `gtk_show_uri()` is fine for GTK+3.

File: src/gtk/assertdlg_gtk.cpp [hunk 13]
```diff
@@ -155,8 +156,8 @@ static void gtk_assert_dialog_save_backtrace_callback(GtkWidget*, GtkAssertDialo
 
     dialog = gtk_file_chooser_dialog_new ("Save assert info to file", GTK_WINDOW(dlg),
                                           GTK_FILE_CHOOSER_ACTION_SAVE,
-                                          GTK_STOCK_CANCEL, GTK_RESPONSE_CANCEL,
-                                          GTK_STOCK_SAVE, GTK_RESPONSE_ACCEPT,
+                                          wxGetTranslation("Cancel").mb_str(), GTK_RESPONSE_CANCEL,
+                                          wxGetTranslation("Save").mb_str(), GTK_RESPONSE_ACCEPT,
```
Comment (paulcor) [inline src/gtk/assertdlg_gtk.cpp]: Should be `utf8_str()`

File: build/aclocal/gtk-4.0.m4 [hunk 14]
```diff
@@ -0,0 +1,208 @@
+# Configure paths for GTK+
```
Comment (paulcor) [inline build/aclocal/gtk-4.0.m4]: Where does `gtk-4.0.m4` come from?

Comment (rkitover) [reply build/aclocal/gtk-4.0.m4]: I made a copy of gtk-3.0.m4 and changed some things.

File: include/wx/gtk/private/addremovectrl.h (hunk 2 above)
Comment (rkitover) [reply include/wx/gtk/private/addremovectrl.h]: I'll just drop this change for now and leave a comment, someone else can look at it when you get there, this is from a different compilation path.

**Comment by rkitover** [comment]
@vadz @paulcor I've made the requested changes.

**Comment by rkitover** [comment]
Oops, forgot the utf8_str() change, pushed.

**Comment by rkitover** [comment]
Travis build should be fine now, was a wxString + GTK2 varargs issue, it's just waiting to do the OS X build now.
